crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2704,7 +2704,80 @@ impl<'db> TypeInferenceBuilder<'db> {
             (_, Type::BytesLiteral(_)) => {
                 self.infer_binary_type_comparison(left, op, KnownClass::Bytes.to_instance(self.db))
             }
+            (Type::Tuple(lhs), Type::Tuple(rhs)) => {
+                // Note: This only works on heterogeneous tuple types.
+                let lhs_elements = lhs.elements(self.db).as_ref();
+                let rhs_elements = rhs.elements(self.db).as_ref();
 
+                match op {
+                    ast::CmpOp::Eq
+                    | ast::CmpOp::NotEq
+                    | ast::CmpOp::Lt
+                    | ast::CmpOp::LtE
+                    | ast::CmpOp::Gt
+                    | ast::CmpOp::GtE => {
+                        self.infer_lexicographic_type_comparison(lhs_elements, op, rhs_elements)
+                    }
+                    ast::CmpOp::In | ast::CmpOp::NotIn => {
+                        let mut eq_cnt = 0usize;
+                        let mut not_eq_cnt = 0usize;
+
+                        for ty in rhs_elements {
+                            let eq_result = self.infer_binary_type_comparison(
+                                Type::Tuple(lhs),
+                                ast::CmpOp::Eq,
+                                *ty,
+                            );
+                            // In Python, if `__eq__` is not defined, it implicitly uses object's `__eq__`. Adding `debug_assert!` as a precaution.
+                            debug_assert!(
+                                eq_result.is_some(),
+                                "Type inference for `==` comparison returned `None`."
+                            );
+
+                            match eq_result {
+                                Some(Type::Todo) | None => return eq_result,
+                                Some(ty) => match ty.bool(self.db) {
+                                    Truthiness::AlwaysTrue => eq_cnt += 1,
+                                    Truthiness::AlwaysFalse => not_eq_cnt += 1,
+                                    Truthiness::Ambiguous => (),
+                                },
+                            }
```

Comment:
If we rely on the invariant you describe in this comment, maybe we should do something like this? (Same applies for the next `debug_assert!` in this PR)

```suggestion
                            let eq_result = self.infer_binary_type_comparison(
                                Type::Tuple(lhs),
                                ast::CmpOp::Eq,
                                *ty,
                            )
                            .expect("infer_binary_type_comparison should never return None for `CmpOp::Eq`");
                            match eq_result {
                                Type::Todo => return Type::Todo,
                                ty => match ty.bool(self.db) {
                                    Truthiness::AlwaysTrue => eq_cnt += 1,
                                    Truthiness::AlwaysFalse => not_eq_cnt += 1,
                                    Truthiness::Ambiguous => (),
                                },
                            }
```